src/v/utils/delta_for.h

```diff
@@ -0,0 +1,1080 @@
+ * - write next 8-bits from every value in a row (16 bytes total);
+ * - write the remaining 3-bits from every value (6 bytes total);
+ * The result is represented using 118 bytes.
+ *
+ * One advantage of this approach is that it requires less code to
+ * implement. It needs bit-packing functions that can pack 1-7 bits
+ * and also 8, 16, 32, and 64 bits. All possible bit-packing arrangements
+ * from 0 to 64 bits can be produced using this functions.
+ * The alternative to this is to implement 63 bit-packing functions that
+ * can pack all possible values. The approach used here requires only
+ * 7 custom bit-packing functions + 4 bit-packing functions for different
+ * size classes (8, 16, 32, 64) which is much easier to impelment and
+ * test.
+ *
+ * It's also beneficial for further compression using general purpose
+ * compression algorithms (e.g. zstd or lz4). If the values have some
+ * common substructure (e.g. the lowest bits are zeroed) this common bits
+ * will be stored together and the compression alg. could take advantage
+ * of that.
+ *
+ * The compressed data is stored internally using an iobuf. This iobuf
+ * can be copied and pushed to the decoder to decompress the values.
+ */
+template<class TVal>
+class deltafor_encoder {
+    static constexpr uint32_t row_width = details::FOR_buffer_depth;
+
+public:
+    explicit deltafor_encoder(TVal initial_value)
+      : _initial(initial_value)
+      , _last(initial_value)
+      , _cnt{0} {}
+
+    deltafor_encoder(
+      TVal initial_value, uint32_t cnt, TVal last_value, iobuf data)
+      : _initial(initial_value)
+      , _last(last_value)
+      , _data(std::move(data))
+      , _cnt(cnt) {}
+
+    using row_t = std::array<TVal, row_width>;
+
+    /// Encode single row
+    void add(const row_t& row) {
+        row_t buf;
+        auto p = _last;
+        uint64_t agg = 0;
+        for (uint32_t i = 0; i < row_width; ++i) {
+            buf[i] = row[i] ^ p;
+            agg |= buf[i];
+            p = row[i];
+        }
+        _last = row.back();
+        uint8_t nbits = 64 - (agg == 0 ? 64 : (uint8_t)__builtin_clzll(agg));
+        _data.append(&nbits, 1);
+        pack(buf, nbits);
+        _cnt++;
+    }
+
+    /// Copy the underlying iobuf
+    iobuf copy() const { return _data.copy(); }
+
+    /// Share the underlying iobuf
+    iobuf share() { return _data.share(0, _data.size_bytes()); }
+
+    /// Return number of rows stored in the underlying iobuf instance
+    uint32_t get_row_count() const noexcept { return _cnt; }
+
+    /// Get initial value used to crate the encoder
+    TVal get_initial_value() const noexcept { return _initial; }
+
+    /// Get last value used to crate the encoder
+    TVal get_last_value() const noexcept { return _last; }
+
+private:
+    template<typename T>
+    void _pack_as(const row_t& input) {
+        static_assert(
+          std::is_unsigned<T>::value,
+          "Only unsigned integer can be used as a type parameter");
```

Comment:
I thought that it's too simple for concept and static_assert will give enough information in case of error. The method is private anyway. But maybe you have something different in mind?

Reply:
Yeah, I agree - it's fine as-is.